### PR TITLE
refactor: pull the refresh period and the service instances discovere…

### PR DESCRIPTION
**Pull request by aureamunoz**
…d cache out. Add tests

### CODE REVIEW

File: service-discovery/kubernetes/src/main/java/io/smallrye/stork/servicediscovery/kubernetes/KubernetesServiceDiscovery.java [hunk 1]
```diff
@@ -44,23 +51,41 @@ public KubernetesServiceDiscovery(String serviceName, ServiceDiscoveryConfig con
...
+        if (refreshNotNeed()) {
+            return Uni.createFrom().item(cacheData.getServiceInstances());
+        }
+        Uni<List<ServiceInstance>> serviceInstances = fetchNewServiceInstancesFromK8s();
+        return serviceInstances.onItem()
+                .invoke(services -> this.cacheData = new ServiceInstancesCache(services, LocalDateTime.now()));
     }
 
-    public Uni<List<ServiceInstance>> getServiceInstances() {
+    @Override
```
Comment (michalszynkiewicz) [inline service-discovery/kubernetes/src/main/java/io/smallrye/stork/servicediscovery/kubernetes/KubernetesServiceDiscovery.java:82]: `refreshNotNeeded` is the same for both, right? Also `getServices`?

File: api/src/main/java/io/smallrye/stork/ServiceInstancesCache.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package io.smallrye.stork;
+
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.util.List;
+
+public class ServiceInstancesCache {
+
+    public static final Duration DEFAULT_REFRESH_INTERVAL = Duration.ofMinutes(5);
+    List<ServiceInstance> serviceInstances;
```
Comment (michalszynkiewicz) [inline api/src/main/java/io/smallrye/stork/ServiceInstancesCache.java:10]: why package-level access?

Comment (michalszynkiewicz) [reply api/src/main/java/io/smallrye/stork/ServiceInstancesCache.java:10]: I was thinking we could either have an abstract superclass for service discovery with caching, or a caching wrapper that would delegate refresh to the actual implementation

File: service-discovery/kubernetes/src/main/java/io/smallrye/stork/servicediscovery/kubernetes/KubernetesServiceDiscovery.java (hunk 1 above)
Comment (aureamunoz) [reply service-discovery/kubernetes/src/main/java/io/smallrye/stork/servicediscovery/kubernetes/KubernetesServiceDiscovery.java:82]: Yes, it's the same. `getServices` also

File: api/src/main/java/io/smallrye/stork/ServiceInstancesCache.java (hunk 2 above)
Comment (aureamunoz) [reply api/src/main/java/io/smallrye/stork/ServiceInstancesCache.java:10]: Yes, I aggree. I was thinking the same at the end but finally commit like currently is. I will give a try this way.
For the visibility of serviceInstances etc, it was just a mistake, I will fix it

File: service-discovery/consul/src/main/java/io/smallrye/stork/servicediscovery/consul/ConsulServiceDiscovery.java [hunk 3]
```diff
@@ -1,38 +1,34 @@
...
 
-    public static final Duration DEFAULT_REFRESH_INTERVAL = Duration.ofMinutes(5);
-    private static final Pattern DIGITS = Pattern.compile("^[-+]?\\d+$");
     private final ConsulClient client;
     private final String serviceName;
-    private final Duration refreshPeriod;
     private boolean passing = true; // default true?
 
-    private volatile ServiceInstancesFetchResult fetchResult;
+    private volatile ServiceInstancesCache cacheData;
```
Comment (michalszynkiewicz) [inline service-discovery/consul/src/main/java/io/smallrye/stork/servicediscovery/consul/ConsulServiceDiscovery.java:31]: is this used?

File: service-discovery/static-list/src/main/java/io/smallrye/stork/servicediscovery/staticlist/StaticListServiceDiscovery.java [hunk 4]
```diff
@@ -4,11 +4,11 @@
 import java.util.List;
 
 import io.smallrye.mutiny.Uni;
+import io.smallrye.stork.CachingServiceDiscovery;
 import io.smallrye.stork.DefaultServiceInstance;
-import io.smallrye.stork.ServiceDiscovery;
 import io.smallrye.stork.ServiceInstance;
 
-public final class StaticListServiceDiscovery implements ServiceDiscovery {
+public final class StaticListServiceDiscovery extends CachingServiceDiscovery {
```
Comment (michalszynkiewicz) [inline service-discovery/static-list/src/main/java/io/smallrye/stork/servicediscovery/staticlist/StaticListServiceDiscovery.java:11]: for this, we can leave it as it was, no need to  cache

Comment (aureamunoz) [reply service-discovery/static-list/src/main/java/io/smallrye/stork/servicediscovery/staticlist/StaticListServiceDiscovery.java:11]: true! you're right

File: service-discovery/consul/src/main/java/io/smallrye/stork/servicediscovery/consul/ConsulServiceDiscovery.java (hunk 3 above)
Comment (aureamunoz) [reply service-discovery/consul/src/main/java/io/smallrye/stork/servicediscovery/consul/ConsulServiceDiscovery.java:31]: oups!

File: service-discovery/kubernetes/src/main/java/io/smallrye/stork/servicediscovery/kubernetes/KubernetesServiceDiscovery.java [hunk 5]
```diff
@@ -44,23 +47,27 @@ public KubernetesServiceDiscovery(String serviceName, ServiceDiscoveryConfig con
         if (namespace != null && namespace.equalsIgnoreCase("all")) {
             allNamespaces = true;
         }
+        String refreshPeriod = config.parameters().get("refresh-period");
```
Comment (michalszynkiewicz) [inline service-discovery/kubernetes/src/main/java/io/smallrye/stork/servicediscovery/kubernetes/KubernetesServiceDiscovery.java:50]: can we read the `refreshPeriod` in the CachingServiceDiscovery class?

Comment (aureamunoz) [reply service-discovery/kubernetes/src/main/java/io/smallrye/stork/servicediscovery/kubernetes/KubernetesServiceDiscovery.java:50]: Let me check

Comment (aureamunoz) [reply service-discovery/kubernetes/src/main/java/io/smallrye/stork/servicediscovery/kubernetes/KubernetesServiceDiscovery.java]: done

**Pull request merged**